src/coreclr/System.Private.CoreLib/src/System/Runtime/DependentHandle.cs

```diff
@@ -0,0 +1,130 @@
+    /// it will do so for the dependent object instance as long as the target is alive.
+    /// <para>
+    /// This type is conceptually equivalent to having a weak reference to a given target object instance A, with
+    /// that object having a field or property (or some other strong reference) to a dependent object instance B.
+    /// </para>
+    /// </summary>
+    public struct DependentHandle : IDisposable
+    {
+        // =========================================================================================
+        // This struct collects all operations on native DependentHandles. The DependentHandle
+        // merely wraps an IntPtr so this struct serves mainly as a "managed typedef."
+        //
+        // DependentHandles exist in one of two states:
+        //
+        //    IsAllocated == false
+        //        No actual handle is allocated underneath. Illegal to get Target, Dependent
+        //        or GetTargetAndDependent(). Ok to call Dispose().
+        //
+        //        Initializing a DependentHandle using the nullary ctor creates a DependentHandle
+        //        that's in the !IsAllocated state.
+        //        (! Right now, we get this guarantee for free because (IntPtr)0 == NULL unmanaged handle.
+        //         ! If that assertion ever becomes false, we'll have to add an _isAllocated field
+        //         ! to compensate.)
+        //
+        //
+        //    IsAllocated == true
+        //        There's a handle allocated underneath. You must call Dispose() on this eventually
+        //        or you cause a native handle table leak.
+        //
+        // This struct intentionally does no self-synchronization. It's up to the caller to
+        // to use DependentHandles in a thread-safe way.
+        // =========================================================================================
+
+        private IntPtr _handle;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DependentHandle"/> struct with the specified arguments.
+        /// </summary>
+        /// <param name="target">The target object instance to track.</param>
+        /// <param name="dependent">The dependent object instance to associate with <paramref name="target"/>.</param>
+        public DependentHandle(object? target, object? dependent)
+        {
+            // no need to check for null result: nInitialize expected to throw OOM.
+            _handle = nInitialize(target, dependent);
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether this handle has been allocated or not.
+        /// </summary>
+        public bool IsAllocated => _handle != IntPtr.Zero;
+
+        /// <summary>
+        /// Gets or sets the target object instance for the current handle.
+        /// </summary>
+        public object? Target
+        {
+            get => nGetPrimary(_handle);
```

Comment:
Does this need null checks like GCHandle? Otherwise, we would crash and burn with hard to understand crash when somebody calls it on nulled out handle.

Reply:
I just left that as the current implementation here:

https://github.com/dotnet/runtime/blob/03b1b38d3d39a60a1e5a27e79a9581a54694a233/src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/DependentHandle.cs#L43

I'd expect the runtime to just throw an exception immediately if someone was to do that, which I guess would be fine? As in, we avoid one extra conditional branch here just to throw maybe a different exception type, but we'd still just throw. Considering the type is very niche and meant for advanced and perf scenarios, it could be fine to leave it as it was? 🤔

The runtime doesn't check this in RELEASE (there's just an `_ASSERTE`):

https://github.com/dotnet/runtime/blob/d0ca735b38a1fe279a05ea0c27370a2f117f5b35/src/coreclr/vm/comdependenthandle.cpp#L57-L64

And indeed (on CoreCLR), trying this out just throw an `ExecutionEngineException`. I'm just thinking that the important bit is getting immediate feedback if this happens (which we do), so considering the scenario we can avoid the overhead to explicitly check this and throw something else? Also because I wouldn't want making this type public to indirectly cause regressions in `ConditionalWeakTable<,>` either, given that that type uses this heavily. 🙂

Reply:
We treat `ExecutionEngineException` as a internal bug in the runtime. It is what the documentation says, it is what automatic classification of crashes does, ... . 

I agree that this type is performance critical, but it is not that performance critical. Each method is still a call.

Reply:
Right, yeah that's fair 🙂
The Mono implementation currently checks as well, and it throws a `NotSupportedException` (which I'm not sure is the right exception type though). What about adding a check to the CoreCLR version to throw `InvalidOperationException`, and then also aligning the Mono version to throw the same type? Otherwise, let me know what you'd prefer!

Reply:
agree

Reply:
Awesome! 😄
Did the changes in 4f602b995ba8f5531136a9370a4af9819e85c194 for both CoreCLR and Mono.
I guess the only remaining questions with respect to usage safety then are:
- Right now `Dispose` is not thread-safe. Should we use `Interlocked.Exchange` to make it so?
- Even if we made `Dispose` thread-safe, the type itself would still not be thread-safe. Eg. if you concurrently called `Dispose` and `Primary`, you could end up trying to retrieving a primary (ie. calling `nGetPrimary`) with a handle that has concurrently been disposed already by another thread. Should we document this?
- The type is not safe at all if someone makes multiple copies (as we discussed in the linked issue with @bartonjs). Do we document this and accept the undefined behavior if someone uses it incorrectly and does weird stuff with multiple instances?

Reply:
Additional question: would you like me to also add some internal unsafe versions of those APIs (without the handle null check) so that `ConditionalWeakTable<,>` would be able to keep using them and not incur in any additional overhead due to these changes? That type had been specifically written from the start to ensure those checks were respected anyway (given that they were not there before), so I'm thinking it might make sense to skip the extra checks in that case? Let me know if you think this would be worth it or not necessary, and I'll make the changes if needed 🙂

Reply:
> Right now Dispose is not thread-safe. Should we use Interlocked.Exchange to make it so?

I do not have an opinion on this one. What do you think?

> Additional question: would you like me to also add some internal unsafe versions of those APIs (without the handle null check

Yes, that's probably a good idea - to avoid dealing with the micro-benchmark regressions.

Reply:
> "I do not have an opinion on this one. What do you think?"

Mmh I'm thinking in this case it might be best to just leave it as is and properly document the whole type as not being thread-safe? Because as mentioned before even if we made `Dispose()` itself thread-safe, the whole type itself would still not be thread-safe, which I guess could potentially be even more confusing for consumers. As in, I wouldn't want some consumers to incorrectly assume that just because `Dispose()` itself is thread-safe, the whole type is as well, potentially leading them towards hard to track bugs. Also, I would expect most people to dispose handles from a finalizer, where thread-safety wouldn't generally be a concern. Does this make sense? 🙂

> "Yes, that's probably a good idea - to avoid dealing with the micro-benchmark regressions."

Great, will do that then! 😄